### ISSUE TITLE
Analytics Opt-Out

### ISSUE THREAD

**Issue by ghost** [freeCodeCamp/devdocs#971]
![screenshot_20190201_115701](https://user-images.githubusercontent.com/440298/52101995-0ca9c000-2619-11e9-86d5-2ce8fa3ebaad.jpg)

ice cat android emui 9

**Comment by j-f1** [comment]
Do you have cookies enabled?

**Comment by j-f1** [comment]
DevDocs needs cookies to store the list of docs you’ve enabled.

**Comment by j-f1** [comment]
I’m not seeing those cookies — are you sure those aren’t coming from an extension? I’ve got a CloudFlare cookie that’s scoped to DevDocs.io, a Google Analytics cookie, some Gauges cookies, and various settings I’ve set in DevDocs.

**Comment by jhabdas** [comment]
Looks like the above snap was coming from the Ledger Manager extension, and activated _specifically_ when visiting DevDocs.io (and not other domains I checked). I have since removed the extension, but I still have a copy if you want to analyze it to determine what it may have been doing to your users.

Here are the cookies coming from your origin, which also include a CloudFlare tracker I'd like to get rid of while still enjoying DevDocs from an IP address in the EU:

<img width="550" alt="screen shot 2019-02-04 at 7 52 00 pm" src="https://user-images.githubusercontent.com/440298/52206889-97f3b180-28b6-11e9-9cbd-8b48dfc4452e.png">

Apart from the trackers the majority of the UX data should probably be moved into Web Storage and the cookies made optional thereafter unless you pop a consent disclaimer.

**Comment by j-f1** [comment]
I believe we store the data in cookies so the server can render the pages and Application Cache manifest properly. If you’re worried about tracking, I’d suggest the DuckDuckGo extension.

**Comment by jhabdas** [comment]
I know this isn't the sexiest subject, but it's been almost a year since GDPR rolled out and I've never seen a cookie disclaimer on DevDocs.io. Perhaps it was overlooked or you're blocking IPs in EU?

https://eugdpr.org/the-regulation/gdpr-faqs/

> **What constitutes personal data?**
> The GDPR applies to ‘personal data’, meaning any information relating to an identifiable person who can be directly or indirectly identified in particular by reference to an identifier. This definition provides for a wide range of personal identifiers to constitute personal data, including name, identification number, location data or online identifier, reflecting changes in technology and the way organisations collect information about people.

Bottom line the tracking ids stored in the DevDocs.io cookies constitute personal data and, therefore, require consent under GDPR legislation for EU.

**Comment by j-f1** [comment]
cc @Thibaut

**Comment by jmerle** [comment]
First of all, IANAL. Everything below here is about the EU.

CloudFlare's `__cfduid` cookie [does not store any personally identifiable information](https://support.cloudflare.com/hc/en-us/articles/200170156-What-does-the-Cloudflare-cfduid-cookie-do-). CloudFlare's `__cfduid` cookie is used to apply security settings on a client-per-client basis and is required for CloudFlare's security features to function properly. The cookie [does not store any personally identifiable information](https://support.cloudflare.com/hc/en-us/articles/200170156-What-does-the-Cloudflare-cfduid-cookie-do-).

The `_ga`, `_gid` and `_gat` cookies are from Google Analytics and are used to distinguish users and throttle requests, thus storing personally identifiable information.

The `_gauges_*` cookies are from Gauges Analytics do not seem to store any personally identifiable information.

The other cookies are used by DevDocs for technical purposes like storing user preferences and do not store any personally identifiable information.

I think there's two legislations we need to adapt to:
- The GDPR, as @jhabdas quoted above, which requires consent for cookies storing personally identifiable information.
- The ePrivacy regulation, which is currently still a proposal but will come in effect somewhere in 2019. This regulation isn't exactly clear to me, but as far as I know it requires consent for all cookies except those that are strictly necessary for the website to function properly. This would mean we'd need consent for both the Google Analytics and the Gauges Analytics cookies, and possible some of DevDocs' own cookies. The CloudFlare cookie would not require consent under this legislation.

DevDocs also has a [privacy policy](https://devdocs.io/about#privacy) which addresses some of the issues mentioned in this GitHub issue. It also states that Google Analytics, Gauges and Sentry are used to collect **anonymous** traffic information, although I do not know if that means that the `_ga` and `_gid` cookies do not store personally identifiable information after all (cc @Thibaut).

I disagree with @j-f1's stand on this issue, we should not require the user to use an extension if they don't want to be tracked.

There are a few solutions to this issue. We could roll our own consent message, but I think it's easier to use a tool like [Cookiebot](https://www.cookiebot.com/en/) (or an alternative) which can add a warning like below:
![](https://support.cookiebot.com/hc/article_attachments/360004739194/Remove_by_Cookiebot_branding.png)

I just submitted a request on Cookiebot's website to check if DevDocs is compliant with GDPR and the ePrivacy Directive (probably not) and checked the box to receive a non-binding offer for Cookiebot to see if we'd fall under the free plan. The free plan is for websites with less than 100 subpages, but I am not sure if they count the content we include dynamically to show the documentation as separate pages. I'll reply again when I receive the results, which could take up to 24 hours.

This would be quite an important change, so it would take some time to implement correctly to ensure compliance with the upcoming legislation.

**Comment by jmerle** [comment]
I received the Cookiebot results. Unfortunately, all the documentation is seen as separate pages, which means DevDocs falls in the most expensive plan (37 euros a month, ouch).

I also read a bit more about the new ePrivacy regulation. Apparently it will overrule the old ePrivacy directive from 2002, which is the directive that requires websites to ask for consent before storing any data not strictly necessary for the website to function properly. The new ePrivacy regulation that will go into effect later this year clarifies that no consent is needed for any non-privacy-intrusive cookies that improve internet experience, or cookies used by a website to count the number of visitors ([source](http://europa.eu/rapid/press-release_IP-17-16_en.htm)).

This would mean that under the new legislation, we would only require consent for the Google Analytics cookies. The Gauge Analytics cookies, the CloudFlare cookie and DevDocs' own cookies do not store personally identifiable information, so they are not privacy-intrusive. I am not exactly sure what is meant by "cookies improving internet experience", so maybe we also need consent for the Gauge Analytics cookies.

To comply with the GDPR, users need to give consent to storing personally identifiable information, so complying with the new ePrivacy regulation would automatically mean complying with the GDPR.

I think we have two categories of cookies on DevDocs:
- Cookies needed for DevDocs to function properly: DevDocs' own cookies and the CloudFlare cookie
- Cookies used for analytical purposes: the Google Analytics and Gauge Analytics cookies

I think it's fair to argue that the first category falls in the "strictly necessary" category, while we would need consent for the second category. Upon first page load we could ask the user if he/she accepts the use of analytical cookies or not, and make it possible to change this consent in the preferences.

If the user does not consent, we can still use Google Analytics by telling it [not to track the user](https://developers.google.com/analytics/devguides/collection/gajs/#disable). By looking at the source code I can see that Gauge Analytics also does tracking, albeit not using cookies. To be on the safe side, I think it's best to disable this too if the user does not consent to analytical cookies.

I'd like @jhabdas opinion on this implementation, as I think he got more knowledge about the matter than my Google Fu could find in a few hours.

**Comment by jmerle** [comment]
I think the implementation suggested in my previous comment is actually better than using Cookiebot, considering it would be more in line with DevDocs' rather unique design.

**Comment by jhabdas** [comment]
First off, I deeply appreciate the attentiveness to this issue. Thank you both for helping get this sorted out quickly. And I'd like to mention up-front that while cookies are a quaffable concept I believe the GDPR regulation applies to Web Storage or any other mechanism which may be used to process and store PII. But that doesn't mean states don't want businesses to collect the data—it's moreso to make individuals are of and the opportunity to give consent _prior_ to the collection of their personal data.

If you want metrics without requesting consent for citizens in the EU you can scrape your logs using a tool like [GoAccess](https://goaccess.io) which I'm personally looking at using after deleting my Google account last May. This is cookie-less and will give you geolocation without necessarily requiring consent unless that data is shared with third parties,  used for purposes other than day to day operations or persisted for what users might consider an unreasonable amount of time.

To touch in the division of cookies @jmerle mentioned... The data CloudFlare collects, along with their current certifications, [may be found](https://www.cloudflare.com/privacyshield/) on their PrivacyShield page. According to Privacy Shield they [are currently](https://www.privacyshield.gov/participant?id=a2zt0000000GnZKAA0) active in both EU-U.S. and Swiss-U.S. with next certifications for each due Nov 2019. And while being an active participant in the Privacy Shield Framework may be necessary for a company [who claims](https://www.cloudflare.com/transparency/) to transmit as much as 10% of all global internet requests, they are indeed **[still collecting personal data](https://www.cloudflare.com/privacypolicy/)** and, for better or worse, [wired in](https://www.cloudflare.com/cookie-policy/) with [surveillance capitalism](https://www.theguardian.com/technology/2019/jan/20/shoshana-zuboff-age-of-surveillance-capitalism-google-facebook). **TL;DR** CloudFlare's cookie is a hybrid and falls into both categories of necessary to function properly and an invasion of personal privacy given the amount of global traffic serviced and the ease with which cookies follow individuals around.

Bottom line, if there's something that can be done to allow the site to function without cookies (I believe there is using Web Storage) the privacy of your users will be better off. If the Manifest requires cookies to function that seems a non-issue as long as the cookie doesn't include personal identifiers of any kind. 🙏

**Comment by jhabdas** [comment]
Just found this thought you might be interested: https://cookieconsent.insites.com/download/

**Comment by jhabdas** [comment]
I noticed the cookies from GitHub [reported earlier](https://github.com/freeCodeCamp/devdocs/issues/971#issuecomment-460216771) are indeed coming from DevDocs. They load when the social widget is loaded in the _Welcome!_ message. And while the welcome message may be turned off clearing the `devdocs.io` `hideIntro` cookie (or cookies in general) brings it right back.

```html
<iframe class="_github-btn" src="//ghbtns.com/github-btn.html?user=freeCodeCamp&amp;repo=devdocs&amp;type=watch&amp;count=true" allowtransparency="true" frameborder="0" scrolling="0" width="100" height="20"></iframe>
```

So that's **two** `__cfuid` tracking cookies to load the welcome page on desktop, **two** `_ga` DoubleClick tracking cookies and a `_gid` tracking cookie from Google; an `_octo` tracking cookie from GitHub which knows your GH username whether you're logged in or not and 5 `_gauges_*` tracking cookies.

So please, don't require cookies to use DevDocs.

**Comment by j-f1** [comment]
#975 removes that iframe and replaces it with an svg.

**Comment by jmerle** [comment]
#979 adds a consent notification so that you have the option to get rid of the Google Analytics and Gauges cookies. The `__cfduid` cookie cannot be deleted without getting rid of CloudFlare altogether, and in my opinion is not a problematic cookie.

I also tried to make DevDocs cookie-less, but turned out to be too big of a change since we use `localStorage` already in order to cache documentation indexes. This cache just grows and grows until `localStorage` is full. Since there is no cross-browser API to check the remaining space in `localStorage`, it's not possible to ensure we leave remaining space for the settings.

**Comment by jhabdas** [comment]
Maybe there's a chance to leverage [Storage Standard](https://storage.spec.whatwg.org/#introduction) later to quash the cookies but opt outs now for the analytics trackers is a marked improvement. Thanks for jumping on this.

**Comment by jhabdas** [comment]
FWIW https://github.com/badges/shields/issues/3026
